### PR TITLE
Metrics comparator

### PR DESCRIPTION
**Pull request by zcain117**
Add utility methods to parse XLA metrics reports into data points and/or generate a report of differences between reports and the historical averages. Include unit tests for main util methods

### CODE REVIEW

**Comment by zcain117** [comment]
> Don't we parse the `Percentiles`?
> We should have a test which grabs the real output of the metrics, and try to parse it, to make sure a change in that part of the code will not break the parser.

I'll add 99 percentile to the set of things we check for each metric.

Is there a way to get a metrics report without running training? If we need to run training, this sounds like it's better as an e2e test of the wrapper rather than having this util function test calling out to model training. What do you think?

**Comment by dlibenzi** [comment]
> > Don't we parse the `Percentiles`?
> > We should have a test which grabs the real output of the metrics, and try to parse it, to make sure a change in that part of the code will not break the parser.
> 
> I'll add 99 percentile to the set of things we check for each metric.

IMHO as part of the parsed context we should capture/parse all the percentiles.
Then the configuration will tel what to cehck.


> Is there a way to get a metrics report without running training? If we need to run training, this sounds like it's better as an e2e test of the wrapper rather than having this util function test calling out to model training. What do you think?

Yeah, just run a new pytorch/xla ops with XLA device, and you have them.

**Comment by zcain117** [comment]
Added logic to parse all the percentiles and added a test to generate a real XLA metrics report and parse it

File: test/metrics_compare_utils_test.py [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+        config={'base_tolerance': 2.0})
+
+    # Since the tolerance is 2.0, the small differences in values are not
+    # big enough to trigger lines in the difference report.
+    expected_report = 'Found new aten counter: aten::_local_scalar_dense__Value: 73216\n'
+    self.assertEqual(metrics_difference_report, expected_report)
+
+  def test_parse_real_metrics(self):
+    print("Testing against TPU. If this hangs, check that $XRT_TPU_CONFIG is set")
+    x = torch.rand(3, 5, device=xm.xla_device())
```
Comment (dlibenzi) [inline test/metrics_compare_utils_test.py]: Please do more ops here, to make sure we capture more metrics (and lower the "surprise rate" 😄 ).

File: torch_xla/debug/metrics_compare_utils.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    else:
+      v_mean = means_and_stddevs[k]['mean']
+      v_stddev = means_and_stddevs[k]['stddev']
+
+      # base_tolerance can be overridden for individual metrics.
+      mul = config.get('{}_tolerance'.format(k), None)
+      if mul is None:
+        mul = config['base_tolerance']
+
+      lower_limit = v_mean - (v_stddev * mul)
```
Comment (dlibenzi) [inline torch_xla/debug/metrics_compare_utils.py]: I think we can do a bit more flexible here.
What we have in the TPU dashboard/monitoring is the ability to specify python expressions, which we then `eval()` with a dictionary built with the metrics inputs.

Also, below, the bounding up/down is not good.
Example, if we improve the step time a lot, should that trigger an alert?

Comment (zcain117) [reply torch_xla/debug/metrics_compare_utils.py]: up/down: Initially I wanted to catch any suspicious activity and also not assume that lower is better for every metric, but I agree that right now lower actually is better for every metric I see.

I removed the lower bound check and added functionality to use a custom expression. That way, if we do end up with a metric where higher is better, we can just pass the expression there. Updated the docstring to explain usage

File: test/metrics_compare_utils_test.py (hunk 1 above)
Comment (zcain117) [reply test/metrics_compare_utils_test.py]: Added a few more and verified that the list of metrics became longer

File: torch_xla/debug/metrics_compare_utils.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+    else:
+      v_mean = means_and_stddevs[k]['mean']
+      v_stddev = means_and_stddevs[k]['stddev']
+
+      # Expression can be overridden for individual metrics.
+      expression = config.get('{}_expression'.format(k))
+      if expression is None:
+        expression = config['base_expression']
+
+      if eval(expression) == False:
```
Comment (dlibenzi) [inline torch_xla/debug/metrics_compare_utils.py]: For security reasons it is better to build a dedicated, limited, dictionary to pass to `eval()`.
Is is also cleaner instead of `eval()` magically picking up from local/global frames.

Comment (zcain117) [reply torch_xla/debug/metrics_compare_utils.py]: Ah yes, great idea. I wasn't familiar with the local/global args for eval()
